storytime

Arinono opened the pull request:
## Authors
@williammartin @Arinono 

Story: [Deploying storytime artifact somewhere for customer stories to run on](https://www.pivotaltracker.com/story/show/178348406)

## Details

Terraform script to deploy storytime artefact.

Arinono commented on storytime/vm.tf:
```diff
@@ -0,0 +1,55 @@
+  version = "~> 2.0"
+
+  container = {
+    image = "docker.io/storyscript/storytime"
```
**Heads up**

The current, deployed, instance for `jem` environment is using an image under `docker.io/arinono/storytime`.

williammartin replied:
We should probably prioritise chores for:
 - Building from commit / master
 - Having a TF var here for the `tag` so we can deploy not only latest.

Arinono commented on storytime/ssl.tf:
```diff
@@ -0,0 +1,26 @@
+resource "google_dns_record_set" "story-pages" {
+  // hardcoded runtime here because I went over the limit of certs generated for
+  // *.storytime.storyscript-ci.com :fullmoon-with-face:
+  name = "${var.app_environment}.runtime.storyscript-ci.com."
```
todo: use the variable

williammartin commented on storytime/.terraform.lock.hcl:
```diff
@@ -0,0 +1,38 @@
+# This file is maintained automatically by "terraform init".
```
Shouldn't be committed

Arinono replied:
Oh, good to know

williammartin replied:
I retracted this, it is important to have.

williammartin commented on storytime/.gitignore:
```diff
@@ -0,0 +1,9 @@
+**/.terraform/*
+
+# .tfstate files
+*.tfstate
```
Good, but let's make sure we're not using local tfstate anyway for Jem/other environments.

williammartin commented on storytime/README.md:
```diff
@@ -0,0 +1,29 @@
+# Deploying GCP VMs located in a private subnet, inside an unmanaged instance group, with a load balancer using Terraform
+
+The script will install two Ubuntu instances with Apache web server located in private subnet, without public ip, inside an unmanaged instance group, with a load balancer using Terraform.
```
No longer correct.

williammartin commented on storytime/variables-auth.tf:
```diff
@@ -0,0 +1,32 @@
+}
+
+# define GCP project name
+variable "app_project" {
```
Not sure "app" is really the right prefix here.

williammartin commented:
Please add an example of the command you need to run to deploy this to the README

JeanBarriere commented on storytime/ssl.tf:
```diff
@@ -0,0 +1,26 @@
+  description = "the cert for the ${var.env_subname}"
+
+  // YOU WILL NEED TO GET THE CERTS ON DISK IF YOU ARE CREATING THIS RESOURCE FROM SCRATCH
+  private_key = file("/path/to/privkey.pem")
+  certificate = file("/path/to/fullchain.pem")
```
**improvements**

maybe use `var.env_something` here?

Arinono replied:
a3e792d

williammartin commented on storytime/lb-unmanaged.tf:
```diff
@@ -0,0 +1,125 @@
+    path_matcher = "storytime"
+  }
+
+  path_matcher {
```
Did you ever figure out why we needed the path rules like this?

Arinono replied:
No, I just kept what was working :/

williammartin replied:
Do you think it is worth understanding?

Arinono replied:
I might be wrong, but I don't think so. Since, at some point, the way the runtime work is going to change. And if we end-up with only one web server (and deal with this properly on the code side), we won't even need path rules at all.

williammartin replied:
Sounds good to me.

Arinono commented:
@williammartin 

I added the TFCloud config to hold the states there.
To be noted: it doesn't deal with applying stuff, only storing the states.

I updated the config and the readme accordingly (to explain how to switch workspaces).

Here is the TFCloud dashboard (to get an example)
![Screen Shot 2021-06-16 at 11 28 42](https://user-images.githubusercontent.com/10957531/122195157-629ffb00-ce96-11eb-8702-e76fa25536cb.png)
![Screen Shot 2021-06-16 at 11 29 46](https://user-images.githubusercontent.com/10957531/122195168-659aeb80-ce96-11eb-9804-49bdd2804271.png)

---

As you can see, I created an (empty) `codecov-ecm` workspace for later, but also for me to see how it's working with the CLI.
The first time I tried to `apply` after `init`, it asked me to pick a workspace to run the config on.
Later, I can switch with `terraform workspace select jem`
